### PR TITLE
[chore] Do regular flush with combined batches

### LINKED ISSUE #22140: [pkg/stanza] Do we need `flushUncombined` in the recombine operator?
### Component(s)

_No response_

### Is your feature request related to a problem? Please describe.

There are a couple of places where we forcefully flush all the batch uncombined:
- when batching reaches `MaxSources` limit
- when `Shutdown` signal is received.

Question: Why this flush has to be uncombined? Can we do the regular flush with combined batches? It likely won't take more time, and we will not need to keep all the entries during the batching, just one (first or latest).

### Describe the solution you'd like

-

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shivanshuraj1333**
fixes: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22140

### CODE REVIEW

File: pkg/stanza/operator/transformer/recombine/recombine.go [hunk 1]
```diff
@@ -315,19 +318,20 @@ func (r *Transformer) addToBatch(_ context.Context, e *entry.Entry, source strin
 // flushUncombined flushes all the logs in the batch individually to the
 // next output in the pipeline. This is only used when there is an error
 // or at shutdown to avoid dropping the logs.
-func (r *Transformer) flushUncombined(ctx context.Context) {
+func (r *Transformer) flushUncombined(ctx context.Context) error {
 	for source := range r.batchMap {
-		for _, entry := range r.batchMap[source].entries {
-			r.Write(ctx, entry)
+		err := r.flushSource(source, true, ctx)
```
Comment (shivanshu1333) [inline pkg/stanza/operator/transformer/recombine/recombine.go:323]: > Why this flush has to be uncombined? Can we do the regular flush with combined batches? It likely won't take more time, and we will not need to keep all the entries during the batching, just one (first or latest).

@dmitryax IIUC this is what you intended to say, correct?

File: pkg/stanza/operator/transformer/recombine/recombine.go [hunk 2]
```diff
@@ -315,19 +318,20 @@ func (r *Transformer) addToBatch(_ context.Context, e *entry.Entry, source strin
 // flushUncombined flushes all the logs in the batch individually to the
 // next output in the pipeline. This is only used when there is an error
 // or at shutdown to avoid dropping the logs.
-func (r *Transformer) flushUncombined(ctx context.Context) {
+func (r *Transformer) flushUncombined(ctx context.Context) error {
```
Comment (dmitryax) [inline pkg/stanza/operator/transformer/recombine/recombine.go:321]: this function doesn't flush _uncombined_ anymore

**Comment by dmitryax** [comment]
The goal of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22140 is to stop keeping the whole `entries` set for every source. This PR can be the first step, but it doesn't resolve the issue

File: pkg/stanza/operator/transformer/recombine/recombine.go [hunk 3]
```diff
@@ -315,19 +318,20 @@ func (r *Transformer) addToBatch(_ context.Context, e *entry.Entry, source strin
...
-		r.removeBatch(source)
 	}
 	r.ticker.Reset(r.forceFlushTimeout)
+	return nil
 }
 
 // flushSource combines the entries currently in the batch into a single entry,
 // then forwards them to the next operator in the pipeline
-func (r *Transformer) flushSource(source string, deleteSource bool) error {
+func (r *Transformer) flushSource(source string, deleteSource bool, ctx context.Context) error {
```
Comment (djaglowski) [inline pkg/stanza/operator/transformer/recombine/recombine.go:334]: This fails lint:
`Error: stanza/operator/transformer/recombine/recombine.go:334:69: context-as-argument: context.Context should be the first parameter of a function (revive)`

File: pkg/stanza/operator/transformer/recombine/recombine.go [hunk 4]
```diff
@@ -305,7 +308,7 @@ func (r *Transformer) addToBatch(_ context.Context, e *entry.Entry, source strin
 	batch.recombined.WriteString(s)
 
 	if (r.maxLogSize > 0 && int64(batch.recombined.Len()) > r.maxLogSize) || len(batch.entries) >= r.maxBatchSize {
-		if err := r.flushSource(source, false); err != nil {
+		if err := r.flushSource(source, false, nil); err != nil {
```
Comment (djaglowski) [inline pkg/stanza/operator/transformer/recombine/recombine.go:311]: Why pass nil for a context rather than `context.Background()`?

**Comment by djaglowski** [comment]
@shivanshu1333, are you planning to pick this one up again?

**Pull request closed without merging**